app/lib/Actions.js

```diff
@@ -0,0 +1,368 @@
+  // If no record to match to is specified (i.e. we are doing a text search), the condition will remain empty for now
+  // Otherwise, it is that a record must match the specified one
+  var match_condition = { ...matchRecord } || {};
+
+  // If the condition contains a component UUID, set it to binary format
+  if (match_condition.componentUuid) {
+    match_condition.componentUuid = MUUID.from(match_condition.componentUuid)
+  }
+
+  // If we are doing a text search, set the 'text' field of the condition
+  if (textSearch) {
+    match_condition['$text'] = { $search: textSearch };
+  }
+
+  // Set up the 'aggregation stages' of the query - these are the query steps in sequence
+  var aggregation_stages = [];
+
+  aggregation_stages.push({ $match: match_condition });
+
+  if (textSearch) {
+    aggregation_stages.push({ $sort: { score: { $meta: 'textScore' } } });
+  }
+
+  aggregation_stages.push({ $sort: { 'insertion.insertDate': -1 } });
+  aggregation_stages.push({ $skip: skip });
+  aggregation_stages.push({ $limit: limit });
+
+  // Query the 'actions' records collection for records matching the condition and additionally defined aggregation stages
+  var records = await db.collection('actions')
+    .aggregate(aggregation_stages)
+    .toArray();
+
+  // Convert the 'componentUuid' of each matching record from binary to string format, for better readability
+  // Additionally, add a 'route' field to each record, which is the URL of that record's information page
+  // Make these changes on copies of the retrieved records, so that the records themselves are not changed
+  var searchResults = [];
+
+  for (var record of records) {
+    var result = { ...record };
+    result.componentUuid = MUUID.from(record.componentUuid).toString();
+    result.route = '/action/' + result.actionId.toString();
+
+    searchResults.push(result);
+  }
+
+  // Return the finalised search results
+  return searchResults;
+}
+
+
+/// Auto-complete an action ID string as it is being typed
+/// This actually returns the records of all actions with a matching action ID to that being typed
+async function autoCompleteId(inputString, typeFormId, limit) {
+  // Set the maximum number of matching action IDs to display
+  limit = limit || 10;
+
+  // The action ID is 24 alphanumeric characters long, so pad the input string out to this length
+  // Then set up objects representing the minimum and maximum binary values that are possible for the current input string
+  var q = inputString.replace(/[_-]/g, '');
+
+  var bitlow = ObjectID(q.padEnd(24, '0'));
+  var bithigh = ObjectID(q.padEnd(24, 'F'));
+
+  // Construct a 'match_condition' to be used as the database query
+  // For this function, it is that the action ID's binary value is between the minimum and maximum binary values defined above
+  var match_condition = {
+    actionId: {
+      $gte: bitlow,
+      $lte: bithigh
+    },
+    typeFormId: null
```

Comment:
@micchickenburger I think you're probably right, but I'm not sure what the correct way is.  I do need to declare `typeformId` as a field in the `match_condition` before the related `if` statement immediately below (otherwise it gives an error saying that it cannot find `match_condition.typeFormId`, which makes sense of course).  So should this line just be `typeFormId = {}`?

Reply:
I don’t think that’s right.  It’s a JavaScript object so members can be added any time in the future.